[Ang Song Yi] iP

songyiang opened the pull request.

Commit:
Add Gradle support

Commit:
Edit Duke's greeting

Commit:
Implement echo and exit features

Commit:
Implement add and list features

Commit:
Implement the ability to mark tasks as done

Commit:
Encapsulate the printing of a task

Commit:
Implement todo, event and deadline features

Commit:
Add comments to Java classes

Commit:
Encapsulate the processing of inputs

Commit:
Handle invalid user inputs

Commit:
Implement delete feature

Commit:
Create duke.txt when user first launches Duke

Commit:
Save tasks' data to duke.txt

Commit:
Store event and deadline dates and times as LocalDateTime

Commit:
Allow user to get tasks from a certain date

Commit:
Load data from duke.txt when Duke starts up

Commit:
Merge branch 'branch-Level-7' into master

Commit:
Merge branch 'branch-Level-8' into master

Commit:
Use more OOP in the code

Commit:
Divide Java classes into packages

Commit:
Merge branch 'add-gradle-support' of into master

Commit:
Add JUnit tests for Parser and DateTimeHandler classes

Commit:
Configure build.gradle to create JAR file

Commit:
Add comments to Java code

Add comments to all field members, constructors and getter and setter
methods.

Commit:
Comply with given coding standard

Commit:
Implement find feature

Commit:
Merge branch 'branch-A-JavaDoc' into master

Commit:
Merge branch 'branch-A-CodingStandard' into master

Commit:
Merge branch 'branch-Level-9' into master

kevnw commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,63 @@
+    /**
+     * Runs Duke.
+     */
+    public void run() {
```
Perhaps can consider changing the method name to become more descriptive. (e.g. runChatbot() or runApplication()).

songyiang replied:
Changed to runDuke().

kevnw commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,236 @@
+     * @param input Input string.
+     * @throws DukeException If input is invalid.
+     */
+    public void processInput(String input) throws DukeException {
```
Perhaps can consider changing the variable input to something like userCommand.

songyiang replied:
Changed to userInput.

kevnw commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,236 @@
+     *
+     * @return Footer string containing the number of tasks.
+     */
+    private String numOfTasksFooter() {
```
Perhaps can consider changing the method name into getNumOfTasks().

songyiang replied:
Changed to getNumOfTaskFooter() to get the footer message when printing out messages.

kevnw commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,159 @@
+     * If it exists, load the data into Duke.
+     * If it does not exist, create one.
+     */
+    public void checkSaveFile() {
```
Perhaps can consider changing the method name into checkSavedFile().

songyiang replied:
Done. Thanks!

kevnw commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,159 @@
+            String line = reader.readLine();
+            while (line != null) {
+                String[] arr = line.split(" / ");
+                Task task;
```
Consider initializing the task variable first.

songyiang replied:
Done. Thanks.

kevnw commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,67 @@
+     * @param i Index of task stored in the list.
+     * @return Task from index i of the list.
+     */
+    public Task getTask(int i) {
```
Perhaps consider changing the variable i into index to be more descriptive.

songyiang replied:
Changed to index.

kevnw commented on src/main/java/duke/task/Event.java:
```diff
@@ -0,0 +1,56 @@
+     * @param at Date and time of the event.
+     * @throws DukeException If the date and time input is invalid.
+     */
+    public Event(String description, String at) throws DukeException {
```
Consider changing the variable at into something more descriptive.

songyiang replied:
Changed from at to dateTime.

kevnw commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,77 @@
+     * @return Status icon.
+     */
+    public String getStatusIcon() {
+        return (this.isDone ? "\u2713" : "\u2718");
```
Good job for using the UTF representation!

songyiang replied:
Thanks :)

kevnw commented on src/test/java/duke/util/DateTimeHandlerTest.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    @Test
+    public void dateFormat_invalidInput_dukeExceptionThrown() {
```
Perhaps can explain more clearly what does it mean by having an invalidInput?

songyiang replied:
Changed to dateFormat_invalidFormat_dukeExceptionThrown().

kevnw reviewed (commented):
Overall, good job! Code is well structured and OOP principles are adhered to really well. Coding style is also aligned with the coding standard. 
Just a few minor nitpicks but the functionality looks good!

WangZijun97 commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,16 @@
+/**
+ * Encapsulates the exception when a parser receives an invalid input.
+ */
+public class DukeException extends Exception {
```
If this is meant to only be thrown when an invalid input is given, perhaps a more descriptive name such as `DukeInputException` would help readability?

songyiang replied:
Done. Thanks.

WangZijun97 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,236 @@
+            this.ui.printMessage("Bye. Hope to see you again soon!");
+        } else if (input.equals("list")) {
+            this.ui.printTaskList(this.taskList);
+        } else if (input.contains("delete")) {
```
Not sure if I'm supposed to give this type of comments, but I think if your user input is something like `todo delete files` your parser will run the wrong command :(

songyiang replied:
Will fix the possibility of user running two or more of such commands stated in this method.

WangZijun97 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,236 @@
+     */
+    private String numOfTasksFooter() {
+        int numOfTasks = this.taskList.getNumOfTasks();
+        if (numOfTasks == 1) {
+            return "Now you have " + numOfTasks + " task in the list.\n";
+        } else {
+            return "Now you have " + numOfTasks + " tasks in the list.\n";
+        }
```
Good attention to detail to differentiate these responses :D

songyiang replied:
Thanks!

WangZijun97 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,159 @@
+    /**
+     * Creates duke.txt in the data directory of the root folder.
+     */
+    private void createTextFile() {
```
Perhaps call these functions `doSomethingSaveFile` rather than `doSomethingTextFile` since you are referring to the save file all along? Could reduce some confusion.

songyiang replied:
In this case, I have to create a directory and a text file. The checkedSaveFile method checks for the existence of both the directory and save file. I think it  will be better to separate these differences. Moreover, the method is private so user will not need to interact with this method.

WangZijun97 commented on src/main/java/duke/task/Deadline.java:
```diff
@@ -0,0 +1,56 @@
+     * @param by Due date and time of the deadline.
+     * @throws DukeException If the date and time input is invalid.
+     */
+    public Deadline(String description, String by) throws DukeException {
```
Perhaps a more descriptive term than `by` can be used to describe the deadline.

songyiang replied:
Changed to dateTime.

WangZijun97 commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,77 @@
+     *
+     * @return Status icon.
+     */
+    public String getStatusIcon() {
```
Should this be `protected`? Not sure if used anywhere outside of `getData()` or `toString()`.

songyiang replied:
Changed to protected.

WangZijun97 reviewed (commented):
Looks clean in general in terms of styling and naming, just some minor nitpicks. As for the program, might want to consider splitting `Parser` into its constituent components, it seems to be doing a lot of things beyond parsing.

Commit:
Edit code style to better adhere to coding standard

Commit:
Encapsulate execution of commands to Executor class

Add Command enum class to represent all possible commands.
This allows code related to the execution of command to be abstracted
out into a new Executor class.

Change test cases in ParserTest to reflect this abstraction.

Commit:
Use Checkstyle to resolve coding style violations

Commit:
Implement GUI with JavaFX

Commit:
Use FXML for GUI design

Commit:
Integrate Duke logic into GUI

Commit:
Allow GUI to read and write data to storage

Commit:
Merge branch 'branch-Level-10' into master

Commit:
Edit code style to better adhere to coding standard

Commit:
Configure build.gradle to create JAR file with GUI

Commit:
Add assertions to Parser class

Commit:
Improve code quality

Use named constants to avoid magic literals for message strings.

Commit:
Merge pull request #1 from SONGYI98/branch-A-Assertions

Add assertions to Parser class

Commit:
Merge pull request #2 from SONGYI98/branch-A-CodeQuality

Improve code quality

Commit:
Use lambda to check task's index in Executor class

Check if the task's index is outside the task list's size range.

Commit:
Use streams for filtering tasks in Executor class

Streams are used to check if the task's date matches the required date
and if the task's description matches the relevant keyword.

Commit:
Set up continuous integration

Add gradle.yml.
Reset Duke's data for I/O redirection test.

Commit:
Edit gradle.yml

Perform I/O redirection test on Windows build only.

Commit:
Repackage Java classes

Shift Storage and TaskList classes into existing util package.
Shift Executor and Parser classes into new logic package.

Commit:
Allow user to attach priority to a task

Users have to attach low, medium or high priority to a task.

Commit:
Implement the ability to filter tasks by priority

Commit:
Improve GUI design

Commit:
Add README.md

Commit:
Display introduction message upon launching GUI

Commit:
Fix GUI using different storage path

CLI and GUI are now using the same storage path to store data.

Commit:
Fix dialog box not expanding when text overrun

Dialog box now expands vertically when there is text overrun.

Commit:
Fix missing error for wrong formats of adding task

An error is now shown to prompt the user to use the correct format when
adding a task.

Commit:
Edit build.gradle

Include JavaFX libraries for Mac and Linux.
